File: src/main/java/de/klierlinge/clear19/widgets/geometry/Rectangle.java
package de.klierlinge.clear19.widgets.geometry;

public final class Rectangle
{
    public static final Rectangle ZEROS = new Rectangle(0, 0, 0, 0, Anchor.TOP_LEFT);
    
    private final int x;
    private final int y;
    private final int width;
    private final int height;

    public Rectangle(int x, int y, int width, int height, Anchor anchor)
    {
        this.width = width;
        this.height = height;

        switch(anchor)
        {
            case TOP_LEFT -> 
            {
                this.x = x;
                this.y = y;
            }
            case TOP_CENTER -> 
            {
                this.x = x - width / 2;
                this.y = y;
            }
            case TOP_RIGHT -> 
            {
                this.x = x - width;
                this.y = y;
            }
            case CENTER_LEFT -> 
            {
                this.x = x;
                this.y = y - height / 2;
            }
            case CENTER_CENTER ->
            {
                this.x = x - width / 2;
                this.y = y - height / 2;
            }
            case CENTER_RIGHT -> 
            {
                this.x = x - width;
                this.y = y - height / 2;
            }
            case BOTTOM_LEFT -> 
            {
                this.x = x;
                this.y = y - height;
            }
            case BOTTOM_CENTER -> 
            {
                this.x = x - width / 2;
                this.y = y - height;
            }
            case BOTTOM_RIGHT -> 
            {
                this.x = x - width;
                this.y = y - height;
            }
            default ->
            {
                this.x = 0;
                this.y = 0;
            }
        }
    }
    
    public Rectangle(AnchoredPoint pos, Size size)
    {
        this(pos.getX(), pos.getY(), size.getWidth(), size.getHeight(), pos.getAnchor());
    }

    public Rectangle(AnchoredPoint from, Vector to)
    {
        this(from, new Size(from, to));
    }
    
    public Size getSize()
    {
        return new Size(width, height);
    }

    public AnchoredPoint getPosition(Anchor anchor)
    {
        return switch(anchor)
        {
            case TOP_LEFT -> new AnchoredPoint(x, y, anchor);
            case TOP_CENTER -> new AnchoredPoint(x + width / 2, y, anchor);
            case TOP_RIGHT -> new AnchoredPoint(x + width, y, anchor);
            case CENTER_LEFT -> new AnchoredPoint(x, y + height / 2, anchor);
            case CENTER_CENTER -> new AnchoredPoint(x + width / 2, y + height / 2, anchor);
            case CENTER_RIGHT -> new AnchoredPoint(x + width, y + height / 2, anchor);
            case BOTTOM_LEFT -> new AnchoredPoint(x, y + height, anchor);
            case BOTTOM_CENTER -> new AnchoredPoint(x + width / 2, y + height, anchor);
            case BOTTOM_RIGHT -> new AnchoredPoint(x + width, y + height, anchor);
        };
    }

    public Rectangle withPosition(AnchoredPoint pos)
    {
        return withPosition(pos.getX(), pos.getY(), pos.getAnchor());
    }
    
    public Rectangle withPosition(int newX, int newY, Anchor anchor)
    {
        return new Rectangle(newX, newY, width, height, anchor);
    }
    
    public Rectangle moved(Vector vector)
    {
        return new Rectangle(getLeft() + vector.getX(), getTop() + vector.getY(), getWidth(), getHeight(), Anchor.TOP_LEFT);
    }
    
    public Rectangle withSize(Size newSize, Anchor anchor)
    {
        return new Rectangle(getPosition(anchor).anchored(anchor), newSize);
    }
    
    public Rectangle withHeight(int newHeight, AnchorH anchor)
    {
        return new Rectangle(getLeft(), getTop(), getWidth(), newHeight, anchor.with(AnchorV.LEFT));
    }
    
    public Rectangle withWidth(int newWidth, AnchorV anchor)
    {
        return new Rectangle(getLeft(), getTop(), newWidth, getHeight(), anchor.with(AnchorH.TOP));
    }
    
    public int getLeft()
    {
        return x;
    }
    
    public int getRight()
    {
        return x + width;
    }
    
    public int getTop()
    {
        return y;
    }
    
    public int getBottom()
    {
        return y + height;
    }
    
    public int getWidth()
    {
        return width;
    }
    
    public int getHeight()
    {
        return height;
    }
    
    @Override
    public int hashCode()
    {
        final int prime = 31;
        int result = 1;
        result = prime * result + height;
        result = prime * result + width;
        result = prime * result + x;
        result = prime * result + y;
        return result;
    }

    @Override
    public boolean equals(Object obj)
    {
        if(this == obj)
            return true;
        if(obj == null)
            return false;
        if(getClass() != obj.getClass())
            return false;
        Rectangle other = (Rectangle)obj;
        if(height != other.height)
            return false;
        if(width != other.width)
            return false;
        if(x != other.x)
            return false;
        if(y != other.y)
            return false;
        return true;
    }

    @Override
    public String toString()
    {
        return "Position [x=" + x + ", y=" + y + ", width=" + width + ", height=" + height + "]";
    }
}

File: clear19/widgets/geometry/rectangle.py
from clear19.widgets.geometry import size
from clear19.widgets.geometry.anchor import Anchor
from clear19.widgets.geometry.point import AnchoredPoint, ZERO_TOP_LEFT
from clear19.widgets.geometry.point import Point
from clear19.widgets.geometry.size import Size


class Rectangle:
    _top_left: Point
    _size: Size

    def __init__(self, point: AnchoredPoint, size: Size):
        self._size = size
        if point.anchor == Anchor.TOP_LEFT:
            self._top_left = Point(point.x, point.y)
        elif point.anchor == Anchor.TOP_CENTER:
            self._top_left = Point(point.x - size.width / 2, point.y)
        elif point.anchor == Anchor.TOP_RIGHT:
            self._top_left = Point(point.x - size.width, point.y)
        elif point.anchor == Anchor.CENTER_LEFT:
            self._top_left = Point(point.x, point.y - size.height / 2)
        elif point.anchor == Anchor.CENTER_CENTER:
            self._top_left = Point(point.x - size.width / 2, point.y - size.height / 2)
        elif point.anchor == Anchor.CENTER_RIGHT:
            self._top_left = Point(point.x - size.width, point.y - size.height / 2)
        elif point.anchor == Anchor.BOTTOM_LEFT:
            self._top_left = Point(point.x, point.y - size.height)
        elif point.anchor == Anchor.BOTTOM_CENTER:
            self._top_left = Point(point.x - size.width / 2, point.y - size.height)
        elif point.anchor == Anchor.BOTTOM_RIGHT:
            self._top_left = Point(point.x - size.width, point.y - size.height)

    @property
    def size(self) -> Size:
        return self._size

    def position(self, anchor: Anchor) -> AnchoredPoint:
        if anchor == Anchor.TOP_LEFT:
            return AnchoredPoint(self._top_left.x, self._top_left.y, anchor)
        elif anchor == Anchor.TOP_CENTER:
            return AnchoredPoint(self._top_left.x + self._size.width / 2, self._top_left.y, anchor)
        elif anchor == Anchor.TOP_RIGHT:
            return AnchoredPoint(self._top_left.x + self._size.width, self._top_left.y, anchor)
        elif anchor == Anchor.CENTER_LEFT:
            return AnchoredPoint(self._top_left.x, self._top_left.y + self._size.height / 2, anchor)
        elif anchor == Anchor.CENTER_CENTER:
            return AnchoredPoint(self._top_left.x + self._size.width / 2, self._top_left.y + self._size.height / 2,
                                 anchor)
        elif anchor == Anchor.CENTER_RIGHT:
            return AnchoredPoint(self._top_left.x + self._size.width, self._top_left.y + self._size.height / 2,
                                 anchor)
        elif anchor == Anchor.BOTTOM_LEFT:
            return AnchoredPoint(self._top_left.x, self._top_left.y + self._size.height, anchor)
        elif anchor == Anchor.BOTTOM_CENTER:
            return AnchoredPoint(self._top_left.x + self._size.width / 2, self._top_left.y + self._size.height,
                                 anchor)
        elif anchor == Anchor.BOTTOM_RIGHT:
            return AnchoredPoint(self._top_left.x + self._size.width, self._top_left.y + self._size.height, anchor)

    def __str__(self) -> str:
        return "Rectangle(Top-Left={}, Size={})".format(self.position(Anchor.TOP_LEFT), self.size)


ZERO: Rectangle = Rectangle(ZERO_TOP_LEFT, size.ZERO)
